server/go/subscription/manager.go

```diff
@@ -0,0 +1,67 @@
+package subscription
+
+import "sync"
+
+type Manager struct {
+	subscriptions map[string][]Subscriber
+	mutex         sync.Mutex
+}
+
+var managerInstance *Manager = nil
+
+func GetManager() *Manager {
+	if managerInstance == nil {
+		managerInstance = newManager()
+	}
+
+	return managerInstance
+}
+
+func newManager() *Manager {
+	return &Manager{
+		subscriptions: make(map[string][]Subscriber),
+		mutex:         sync.Mutex{},
+	}
+}
+
+func (m *Manager) Subscribe(resourceID string, subscriber Subscriber) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	array := make([]Subscriber, 0)
+
+	if existingArray, ok := m.subscriptions[resourceID]; ok {
+		array = existingArray
+	}
+
+	array = append(array, subscriber)
+	m.subscriptions[resourceID] = array
+}
+
+func (m *Manager) Unsubscribe(resourceID string, subscriptionID string) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if array, ok := m.subscriptions[resourceID]; ok {
+		newArray := make([]Subscriber, 0, len(array)-1)
+		for _, item := range array {
+			if item.ID() != subscriptionID {
+				newArray = append(newArray, item)
+			}
+		}
+
+		m.subscriptions[resourceID] = newArray
+	}
+
+}
+
+func (m *Manager) PublishUpdate(resourceID string, message Message) {
+	m.mutex.Lock()
```

Comment:
why is this lock needed? I _think_ that since this func doesn't write anything, it can work without lock

Reply:
If you read and write to a map concurrently, golang might throw this error:

```
fatal error: concurrent map read and map write
```

You can [try it here](https://go.dev/play/p/TIZHgEpME_2)

Reply:
Ahh ok. ignore this then!